### ISSUE TITLE
add getRootGraph and isLineGraph to LineGraphConverter

### ISSUE THREAD

**Issue by jsichi** [jgrapht/jgrapht#690]
Followup item from #687 

```
//    To implement getRootGraph and isLineGraph, see:
//    -Lehot, P. G. H. "An Optimal Algorithm to Detect a Line Graph and Output Its Root Graph." J. ACM 21, 569-575, 1974.
//    -Sysło, Maciej M. (1982), "A labeling algorithm to recognize a line digraph and output its root graph", Information Processing Letters, 15 (1): 28–30, doi:10.1016/0020-0190(82)90080-1, MR 0678028.
//    -Degiorgi D.G., Simon K. (1995) A dynamic algorithm for line graph recognition. In: Nagl M. (eds) Graph-Theoretic Concepts in Computer Science. WG 1995. Lecture Notes in Computer Science, vol 1017. Springer, Berlin, Heidelberg
//    public void convertToRootGraph(Graph<V, E> target){
//        //Not yet implemented
//    }
//
//    Determines whether the graph is a line graph
//    public boolean isLineGraph(){
//        //Not yet implemented. Add shortcut to GraphTest
//    }
}
```

**Comment by Gaurav-Punjabi** [comment]
Can I Solve this issue ?

**Comment by jkinable** [comment]
Sure, feel free to implement this and submit a PR when you are ready. Make sure you read the [dev wiki](https://github.com/jgrapht/jgrapht/wiki#developer-pages) first.

**Comment by Gaurav-Punjabi** [comment]
Okay will do that

**Comment by jsichi** [comment]
Yeah, that's a good point...for the inverse operation, the generic parameter types don't match the forward direction.

@jkinable what do you think?

**Comment by Gaurav-Punjabi** [comment]
@jkinable @jsichi So should I create a new class named RootGraphConverter and add the 2 functionalities in it?

**Comment by jsichi** [comment]
Option 2 is fine with me, assuming javac is happy with it :)

@Gaurav-Punjabi give option 2 a try and if you don't run into any trouble from the generics when you write your unit tests, we'll go with that.

**Comment by Gaurav-Punjabi** [comment]
Okay that's fine.

**Comment by Gaurav-Punjabi** [comment]
@jkinable @jsichi Hey I have implemented the method for convertToRootGraph() with the help of option 2.
But for checking whether the given graph is a line graph or not. The steps are : 
1 - Convert the line graph to root graph
2- Convert the resulting root graph to line graph.
3 - Compare the given line graph and the converted line graph. If they are equal then the given graph is line graph else it is not a line graph.
**Issue**
So the issue is for converting the given line graph to root graph I need to create a local Graph<U, V>.
But for doing that I need the class instance of V as required by the Graph's constructor.
So what should I do? Should I take class instance as a parameter in the constructor or what?

**Comment by jkinable** [comment]
@Gaurav-Punjabi which paper are you using for the test whether a given graph is a line graph?

**Comment by jkinable** [comment]
Btw, in retrospect, I think option (1) is better than option (2).
```
LineGraphConverter<V,E,U>{
    private final Graph<V,E> graph;
    public LineGraphConverter(Graph<V,E> graph){
        this.graph=graph;
    }

    public void convertToLineGraph(final Graph<E, U> target){
        //Do something
    }

    public void convertToRootGraph(final Graph<U, V> target){
        //Do something
    }

    public void isLineGraph(){
        //Do something
    }
}
```
If the user just wants to know whether the given input graph is a line graph, then it doesn't seem to make sense to initialize the class with a generic parameter U. So instead of the above, you can try this:

```
LineGraphConverter<V,E>{

    private final Graph<V,E> graph;

    public LineGraphConverter(Graph<V,E> graph){
        this.graph=graph;
    }

    public <EE> void convertToLineGraph(final Graph<E, EE> target){
        //Do something
    }

    public <VV> void convertToRootGraph(final Graph<VV, V> target){
        //Do something
    }

    public void isLineGraph(){
        //Do something
    }
}
```
This should make the implementation of isLineGraph() relatively straightforward.

**Comment by Gaurav-Punjabi** [comment]
@jkinable I'm using "A dynamic algorithm for line graph recognition" By Daniele Giorgio Degiorgi, Klaus Simon

**Comment by Gaurav-Punjabi** [comment]
@jkinable I have changed my code and implemented option 1, but I'm still stuck on the conversion problem tell me what should I do? Should I take the class instance as a parameter in the constructor?

**Comment by jkinable** [comment]
@Gaurav-Punjabi I'm not sure I understand the issue. Why would you need the class instance parameter in the constructor? If you use this, I don't see why you would want to pass the class instance as a parameter in the constructor:
```
LineGraphConverter<V,E>{

    private final Graph<V,E> graph;

    public LineGraphConverter(Graph<V,E> graph){
        this.graph=graph;
    }

    public <EE> void convertToLineGraph(final Graph<E, EE> target){
        //Do something
    }

    public <VV> void convertToRootGraph(final Graph<VV, V> target){
        //Do something
    }

    public void isLineGraph(){
        //Do something
    }
}
```

It would help if you refer to code such that I can see what the problem is. I checked your [code](https://github.com/Gaurav-Punjabi/jgrapht/blob/feature/line-graph/jgrapht-core/src/main/java/org/jgrapht/alg/transform/LineGraphConverter.java), but there the class still has the <V,E,EE> generics instead of <V,E>.

**Comment by Gaurav-Punjabi** [comment]
@jkinable Oh yeah i havent pushed the code yet, wait first ill push it and then explain you the issue with isLineGraphMethod so you can refer the latest code.

**Comment by Gaurav-Punjabi** [comment]
Okay, the issue is with isLineGraph() method.
The algorithm for checking whether the given graph is LineGraph or not is : 
1 - Convert the given graph to RootGraph.
2 - Convert the RootGraph to LineGraph.
3 -  Compare the Resulting LineGraph and given Graph if both are the same then the given graph is a line graph.

So to convert the given graph into root graph I need to create a local graph in isLineGraph() which is not possible as the constructor of Graph requires a class instance of the edge type.
So what should I do? @jkinable

**Comment by jkinable** [comment]
> So which one should i implement ?

The fastest, best performing method.

> So to convert the given graph into root graph I need to create a local graph in isLineGraph() which is not possible as the constructor of Graph requires a class instance of the edge type.
There are various ways you could resolve this. One of them is to create a copy of the input graph using vertex/edge types you know, e.g. `Graph<V,E> graphToBeTested=....; Graph<Integer,DefaultEdge> localCopy=....; `For each vertex and for each edge in graphToBeTested, add corresponding vertex/edge in `localCopy`. For the isLineGraph test you really don't care what type the vertex/edge is.
There are possibly more elegant solutions which do not require a copy.
Part of the exercise if to play around with this and figure out an elegant solution.
